### PR TITLE
Speedup of utilities calculation and simulation

### PR DESCRIPTION
**Pull request by boryana-ilieva**
We reorganised the calculation of utilities and rewrote the simulation module to ensure highest possible speed for the current soepy version.

### CODE REVIEW

**Comment by boryana-ilieva** [comment]
@janosg  & @tobiasraabe, could you please share your thoughts on the first and the second point

**Comment by boryana-ilieva** [comment]
As far as last point is concerned: We needed to change the form in order to be able to regression test the changes we made to speed up the calculation of utilities. Specifically, we needed the ability to test equality only of certain columns and not the entire data frame.
This can potentially also be useful in the future, but we can happily switch back to the checksum if you see an important reason to do so.

**Comment by tobiasraabe** [comment]
I do not think this PR makes much sense. If you want to sync develop and master, just switch to master and perform a fast-forward merge with ``git merge develop --ff-only``. The reason is we want to keep the complete commit history of PRs, but not commits in a PR. Currently, only squash commits are allowed for PRs so we would loose the history. I prefer to keep this restriction as otherwise one accidentally performs a simple merge and the history gets polluted as with the regression tests PR.

Now, to the bullet points.

1. ``.envrc``, the ``.idea`` folder and objects similar to this are actually user dependent and should not be committed to a repository. Such files can be added to ``.gitignore`` and left in the project folder. Adding the project folder to ``PYTHONPATH`` is done by making an editable install of soepy ``pip install -e soepy``. Then, the project folder on the path and ``pip install soepy`` cannot coexist and interfere with each other as it is pip install or pip install -e only but not both.
2. We changed from json to pickle to save the complete dataframe and not only the checksum value.
3. Changing from checksum to dataframe allows us to keep the regression tests in cases where the overall dataframe has changed. This happened while restructuring the rewards. If a dataframe would have been available, we had not lost the vault. Furthermore, it is easier to diagnose where deviations occurred in cases of failures.

**Comment by janosg** [comment]
I agree with @tobiasraabe's points. We chose pickle over a human readable format to save disc space. Alternatively you could use a more readable format and just save some rows of the dataset.

**Comment by peisenha** [comment]
Thanks for the comments, I agree with the advantages of storing (maybe smaller samples) of the whole data frame to ease debugging and have the ability to maybe as intermediate steps only test columns that are supposedly unaffected by a change even though others might change ... and thus result in a differing checksum. I prefer our old development setup, so `.envrc' and the alignment of the Pycharm code inspection are reactivated in master fixup.

File: soepy/python/simulate/simulate_auxiliary.py [hunk 1]
```diff
@@ -11,99 +16,122 @@ def pyth_simulate(model_params, states, indexer, emaxs, covariates):
...
+                np.tile(nonconsumption_utilities, (current_states.shape[0], 1)),
+                continuation_values,
+                value_functions,
+            )
+        )
+
+        # Update current states
+        current_states[:, 1] += 1
+        current_states[:, 3] = choice
+        # current_states[np.arange(current_states.shape[0]), choice + 3] = np.where(
```
Comment (peisenha) [inline soepy/python/simulate/simulate_auxiliary.py]: Can we remove this?

File: soepy/test/test_regression.py [hunk 2]
```diff
@@ -22,5 +22,9 @@ def test1(idx):
 
     calculated_df = simulate(init_dict)
 
-    for col in expected_df.filter(like="Value Functions").columns.tolist():
-        expected_df[col].equals(calculated_df)
+    for col in expected_df.filter(like="Value_Functions").columns.tolist():
+        print(col)
```
Comment (peisenha) [inline soepy/test/test_regression.py]: Please remove.

File: soepy/test/test_regression.py [hunk 3]
```diff
@@ -22,5 +22,9 @@ def test1(idx):
 
     calculated_df = simulate(init_dict)
 
-    for col in expected_df.filter(like="Value Functions").columns.tolist():
-        expected_df[col].equals(calculated_df)
+    for col in expected_df.filter(like="Value_Functions").columns.tolist():
```
Comment (peisenha) [inline soepy/test/test_regression.py]: #69  please address right here

File: soepy/python/simulate/simulate_auxiliary.py (hunk 1 above)
Comment (boryana-ilieva) [reply soepy/python/simulate/simulate_auxiliary.py]: Are you referring to the line which is commented out?

Comment (peisenha) [reply soepy/python/simulate/simulate_auxiliary.py]: Yes

Comment (boryana-ilieva) [reply soepy/python/simulate/simulate_auxiliary.py]: Done

File: soepy/test/test_regression.py (hunk 2 above)
Comment (boryana-ilieva) [reply soepy/test/test_regression.py]: Done.

File: soepy/test/test_regression.py (hunk 3 above)
Comment (boryana-ilieva) [reply soepy/test/test_regression.py]: Done.

**Comment by boryana-ilieva** [comment]
Updated outdated doc strings.

**Pull request merged**